src/zkbob/ZkBobPool.sol

```diff
@@ -242,6 +307,143 @@ contract ZkBobPool is EIP1967Admin, Ownable, Parameters, ZkBobAccounting {
+        require(
+            batch_deposit_verifier.verifyProof(batch_deposit_pub, _batch_deposit_proof),
+            "ZkBobPool: bad batch deposit proof"
+        );
+
+        uint256[3] memory tree_pub = [roots[_pool_index], _root_after, _out_commit];
+        require(tree_verifier.verifyProof(tree_pub, _tree_proof), "ZkBobPool: bad tree proof");
+
+        _pool_index += 128;
+        roots[_pool_index] = _tree_root_after();
+        bytes memory message; // TODO
+        assembly {
+            message := batch_deposit_pub
+            mstore(message, mul(64, calldataload(_indices.offset)))
+        }
+        bytes32 message_hash = keccak256(message);
+        bytes32 _all_messages_hash = keccak256(abi.encodePacked(all_messages_hash, message_hash));
+        all_messages_hash = _all_messages_hash;
+        emit Message(_pool_index, _all_messages_hash, message);
+        emit CompleteDirectDepositBatch(_pool_index, _indices);
+    }
+
+    function directDeposit(address _fallbackUser, uint256 _amount, bytes memory _rawZkAddress) external {
+        _recordDirectDeposit(_fallbackUser, _amount, _rawZkAddress);
+    }
+
+    function onTokenTransfer(address _from, uint256 _value, bytes calldata _data) external returns (bool) {
+        require(msg.sender == token, "ZkBobPool: not a token caller");
+
+        (address fallbackUser, bytes memory rawZkAddress) = abi.decode(_data, (address, bytes));
+
+        _recordDirectDeposit(fallbackUser, _value, rawZkAddress);
+
+        return true;
+    }
+
+    function refundDirectDeposit(uint256 _index) external {
+        bool isOperator = operatorManager.isOperator(msg.sender);
+        DirectDeposit storage dd = directDeposits[_index];
+        require(dd.status == DirectDepositStatus.Pending, "ZkBobPool: direct deposit not pending");
+        require(
+            isOperator || dd.timestamp + directDepositTimeout < block.timestamp,
+            "ZkBobPool: direct deposit timeout not passed"
+        );
+        _refundDirectDeposit(_index, dd);
+    }
+
+    function refundDirectDeposit(uint256[] calldata _indices) external {
+        bool isOperator = operatorManager.isOperator(msg.sender);
+
+        for (uint256 i = 0; i < _indices.length; i++) {
+            DirectDeposit storage dd = directDeposits[_indices[i]];
+
+            if (dd.status == DirectDepositStatus.Pending) {
+                require(
+                    isOperator || dd.timestamp + directDepositTimeout < block.timestamp,
+                    "ZkBobPool: direct deposit timeout not passed"
+                );
+                _refundDirectDeposit(_indices[i], dd);
+            }
+        }
+    }
+
+    function _refundDirectDeposit(uint256 _index, DirectDeposit storage _dd) internal {
+        _dd.status = DirectDepositStatus.Refunded;
+
+        (address user, uint96 amount) = (_dd.user, _dd.amount);
+
+        IERC20(token).safeTransfer(user, amount);
```

Comment:
should we still return entire amount if a refund was initiated by the operator?

Reply:
I didn't have any particular use case idea in mind for operator refunds, but I don't think we should take any fee in that case